File: src/main/java/burp/BurpExtender.java
package burp;

import java.awt.*;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class BurpExtender implements IBurpExtender, IMessageEditorTabFactory {
    private IBurpExtenderCallbacks callbacks;
    private IExtensionHelpers helpers;

    public String ExtenderName = "unidecode";

    @Override
    public void registerExtenderCallbacks(IBurpExtenderCallbacks callbacks) {
        this.callbacks = callbacks;
        helpers = callbacks.getHelpers();
        callbacks.printOutput(ExtenderName + " by 4shen0ne");
        callbacks.setExtensionName(ExtenderName);

        callbacks.registerMessageEditorTabFactory(this);
    }

    @Override
    public IMessageEditorTab createNewInstance(IMessageEditorController controller, boolean editable) {
        return new ChineseResponseTab(controller, helpers);
    }

    class ChineseResponseTab implements IMessageEditorTab {
        private IMessageEditorController controller;
        private IExtensionHelpers helpers;
        private ITextEditor txtInput;

        private Pattern pattern = Pattern.compile("(\\\\u(\\p{XDigit}{4}))");
        private byte[] currentMessage;

        public ChineseResponseTab(IMessageEditorController controller, IExtensionHelpers helpers) {
            this.controller = controller;
            this.helpers = helpers;

            txtInput = callbacks.createTextEditor();
        }

        @Override
        public String getTabCaption() {
            return "unidecode";
        }

        @Override
        public Component getUiComponent() {
            return txtInput.getComponent();
        }

        @Override
        public boolean isEnabled(byte[] content, boolean isRequest) {
            return !isRequest && hasUnicode(content);
        }

        private boolean hasUnicode(byte[] content) {
            return pattern.matcher(new String(content).toLowerCase()).find();
        }

        @Override
        public void setMessage(byte[] content, boolean isRequest) {
            IResponseInfo respInfo = helpers.analyzeResponse(content);
            // only decode response body
            String bodyContent = new String(Arrays.copyOfRange(content, respInfo.getBodyOffset(), content.length));

            txtInput.setText(currentMessage = unidecode(bodyContent));
        }

        private byte[] unidecode(String bodyContent) {
            Matcher matcher = pattern.matcher(bodyContent.toLowerCase());
            StringBuilder result = new StringBuilder();
            int index = 0;
            char decoded;
            while (matcher.find()) {
                // hex decode
                decoded = (char) Integer.parseInt(bodyContent.substring(matcher.start() + 2, matcher.end()), 16);
                result.append(bodyContent.substring(index, matcher.start()) + decoded);
                index = matcher.end();
            }
            result.append(bodyContent.substring(index));
            return String.valueOf(result).getBytes(StandardCharsets.UTF_8);
        }

        @Override
        public byte[] getMessage() {
            return currentMessage;
        }

        @Override
        public boolean isModified() {
            return txtInput.isTextModified();
        }

        @Override
        public byte[] getSelectedData() {
            return txtInput.getSelectedText();
        }
    }
}

File: src/main/kotlin/burp/BurpExtender.kt
package burp

import kotlin.text.Regex
import java.awt.Component

const val EXTENDER_NAME = "unidecode"

class BurpExtender : IBurpExtender {
    private lateinit var callbacks: IBurpExtenderCallbacks

    override fun registerExtenderCallbacks(callbacks: IBurpExtenderCallbacks) {
        this.callbacks = callbacks
        callbacks.printOutput("$EXTENDER_NAME by 4shen0ne")
        callbacks.setExtensionName(EXTENDER_NAME)
        callbacks.registerMessageEditorTabFactory { controller, _ ->
            ChineseResponseTab(controller, callbacks)
        }
    }
}

class ChineseResponseTab(
    controller: IMessageEditorController, callbacks: IBurpExtenderCallbacks
) : IMessageEditorTab {
    private val txtInput: ITextEditor = callbacks.createTextEditor()
    private val helpers: IExtensionHelpers = callbacks.helpers
    private val regex: Regex = Regex("(\\\\u(\\p{XDigit}{4}))")
    private lateinit var currentMessage: ByteArray

    override fun getMessage(): ByteArray = currentMessage
    override fun getTabCaption(): String = EXTENDER_NAME
    override fun isModified(): Boolean = txtInput.isTextModified
    override fun getSelectedData(): ByteArray = txtInput.selectedText
    override fun getUiComponent(): Component = txtInput.component

    override fun isEnabled(
        content: ByteArray, isRequest: Boolean
    ): Boolean = !isRequest && hasUnicode(content)

    private fun hasUnicode(content: ByteArray): Boolean {
        return regex.containsMatchIn(String(content).lowercase())
    }

    override fun setMessage(content: ByteArray, isRequest: Boolean) {
        val respInfo: IResponseInfo = helpers.analyzeResponse(content)
        // only decode response body
        val bodyContent = String(content.copyOfRange(respInfo.bodyOffset, content.size))
        txtInput.text = unidecode(bodyContent).also { currentMessage = it }
    }

    private fun unidecode(bodyContent: String): ByteArray =
        regex.replace(bodyContent.lowercase()) {
            it.value.substring(2).toInt(16).toChar().toString()
        }.toByteArray()
}
